packages/universal-router-sdk/src/entities/protocols/pancakeswap.ts

```diff
@@ -0,0 +1,436 @@
+          ])
+        } else {
+          planner.addCommand(CommandType.SWEEP, [
+            sampleTrade.outputAmount.currency.wrapped.address,
+            this.options.recipient,
+            BigInt(minAmountOut.quotient.toString()),
+          ])
+        }
+      }
+
+      if ((inputIsNative && this.type === TradeType.EXACT_OUTPUT) || SwapRouter.riskOfPartialFill(trades)) {
+        // for exactOutput swaps that take native currency as input
+        // we need to send back the change to the user
+        planner.addCommand(CommandType.UNWRAP_WETH, [this.options.recipient, 0n])
+      }
+    }
+  }
+}
+
+// encode a v2 swap
+function addV2Swap(
+  planner: RoutePlanner,
+  trade: SmartRouterTrade<TradeType>,
+  options: PancakeSwapOptions,
+  routerMustCustody: boolean,
+  payerIsUser: boolean,
+  performAggregatedSlippageCheck: boolean,
+): void {
+  const amountIn = BigInt(SmartRouter.maximumAmountIn(trade, options.slippageTolerance).quotient.toString())
+  const amountOut = BigInt(SmartRouter.minimumAmountOut(trade, options.slippageTolerance).quotient.toString())
+
+  // V2 trade should have only one route
+  const route = trade.routes[0]
+  const path = route.path.map((token) => token.wrapped.address)
+  const recipient = routerMustCustody ? ROUTER_AS_RECIPIENT : validateAndParseAddress(options.recipient!)
+
+  // same as encodeV2Swap only we dont return calldatas instead we push to the command planner
+  if (trade.tradeType == TradeType.EXACT_INPUT) {
+    planner.addCommand(CommandType.V2_SWAP_EXACT_IN, [
+      // if native, we have to unwrap so keep in the router for now
+      recipient,
+      amountIn,
+      performAggregatedSlippageCheck ? 0n : amountOut,
+      path,
+      payerIsUser,
+    ])
+  } else if (trade.tradeType == TradeType.EXACT_OUTPUT) {
+    planner.addCommand(CommandType.V2_SWAP_EXACT_OUT, [recipient, amountOut, amountIn, path, payerIsUser])
+  }
+}
+
+// encode a v3 swap
+async function addV3Swap(
+  planner: RoutePlanner,
+  trade: SmartRouterTrade<TradeType>,
+  options: PancakeSwapOptions,
+  routerMustCustody: boolean,
+  payerIsUser: boolean,
+  performAggregatedSlippageCheck: boolean,
+): Promise<void> {
+  for (const route of trade.routes) {
+    const { inputAmount, outputAmount } = route
+
+    // we need to generaate v3 path as a hash string. we can still use encodeMixedRoute
+    // as a v3 swap is essentially a for of mixedRoute
+    const path = SmartRouter.encodeMixedRouteToPath(
+      { ...route, input: inputAmount.currency, output: outputAmount.currency },
+      trade.tradeType === TradeType.EXACT_OUTPUT,
+    )
+    const amountIn = BigInt(SmartRouter.maximumAmountIn(trade, options.slippageTolerance).quotient.toString())
+    const amountOut = BigInt(SmartRouter.minimumAmountOut(trade, options.slippageTolerance).quotient.toString())
+
+    const recipient = routerMustCustody ? ROUTER_AS_RECIPIENT : validateAndParseAddress(options.recipient!)
```

Comment:
let's try avoid using non empty assertion as well

Reply:
sure will fix this